homeassistant/components/backup/backup.py

```diff
@@ -0,0 +1,146 @@
+"""Local backup support for Core and Container installations."""
+
+from __future__ import annotations
+
+from dataclasses import asdict, dataclass
+import json
+from pathlib import Path
+import tarfile
+from tarfile import TarError
+from typing import Any, cast
+
+from homeassistant.core import HomeAssistant
+from homeassistant.util.json import json_loads_object
+
+from .agent import BackupAgent, UploadedBackup
+from .const import BUF_SIZE, LOGGER
+from .models import BackupUploadMetadata
+
+
+async def async_get_backup_agents(
+    hass: HomeAssistant,
+    **kwargs: Any,
+) -> list[BackupAgent]:
+    """Register the backup agent."""
+    return [LocalBackupAgent(hass)]
+
+
+@dataclass(slots=True)
+class LocalBackup(UploadedBackup):
+    """Local backup class."""
+
+    path: Path
+
+    def as_dict(self) -> dict:
+        """Return a dict representation of this backup."""
+        return {**asdict(self), "path": self.path.as_posix()}
+
+
+class LocalBackupAgent(BackupAgent):
+    """Define the format that backup agents can have."""
+
+    name = "local"
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the backup agent."""
+        super().__init__()
+        self._hass = hass
+        self.backup_dir = Path(hass.config.path("backups"))
+        self.backups: dict[str, LocalBackup] = {}
+        self.loaded_backups = False
+
+    async def load_backups(self) -> None:
+        """Load data of stored backup files."""
+        backups = await self._hass.async_add_executor_job(self._read_backups)
+        LOGGER.debug("Loaded %s local backups", len(backups))
+        self.backups = backups
+        self.loaded_backups = True
+
+    def _read_backups(self) -> dict[str, LocalBackup]:
+        """Read backups from disk."""
+        backups: dict[str, LocalBackup] = {}
+        for backup_path in self.backup_dir.glob("*.tar"):
+            try:
+                with tarfile.open(backup_path, "r:", bufsize=BUF_SIZE) as backup_file:
+                    if data_file := backup_file.extractfile("./backup.json"):
+                        data = json_loads_object(data_file.read())
+                        backup = LocalBackup(
+                            id=cast(str, data["slug"]),  # Do we need another ID?
```

Comment:
The `UploadedBackup` base class says we need an ID, I'm not sure why, manager and frontend should care about the slug only.

Reply:
The id is the agent's way of identifying the backup. The slug is something that the manager uses to identify the backup. They could be the same but I think it's good to separate terms since the semantics are different.

Reply:
OK, but if we don't need the additional identifier there's no point in storing a reference to the slug twice.
The manager will only ever ask the agent to handle a backup based on the slug. The ID is an implementation detail which individual agents can add if they need one.

Reply:
Right now the id is part of the interface and the id parameter is used to eg download a backup. We could remove it and force agents to keep track of backups via the slug.

I don't really like the name slug since it doesn't describe the purpose of the parameter so I'd prefer to limit its use in interfaces.

Reply:
OK.
I suggest we rename it to `backup_id` which makes the purpose clear.
Local agents can then add their own ID if needed for internal purpose.

Reply:
Perfect.